### DISCUSSION TITLE
[Idea] Plugin-structure for qpdf to be used in, for instance, okular, evince, atril and so forth

### DISCUSSION THREAD

**Discussion by rubyFeedback** [qpdf/qpdf#613] [Ideas]
Hey there Jay and qpdf users in general,

This is more an idea, so please feel free to disregard / close at any moment in time.

I am using qpdf via ruby from the commandline and it works very well. I also have a
tiny "wrapper" built via ruby-gtk3 - this is more a hacky patch and not great but I
wanted to add useful GUI-relatable functionality.

This hobby project is nowhere on the level of something such as okular (from
the KDE software suite).

One recent thing I became interested in is annotations:

https://docs.kde.org/trunk5/en/okular/okular/configannotations.html

I read through books in .pdf form and after I have read them, I remove the
.pdf file usually. I also chop off individual pages that I have read. I do make
notes, though, so if need be I simply obtain a new .pdf file to look / check on
the original content, but by and large this strange behaviour is how I tend
to use .pdf files right now. So looking into annotations would make sense,
IMO.

So then I had a second thought ... what if I could read in KDE okular, but
actually remove pages as-is, via qpdf? That is I would use KDE okular
for reading .pdf files but use qpdf for manipulating the .pdf file at hand.
And I could use annotations too e. g. mark where I last read something
on a page in colour or something. (I actually do not always remove 
pages from .pdf files; usually I may work through a chapter and re-read
it some days later  until I fully understand it before removing that
chapter then.)

So ... this then got me thinking ... what if qpdf would allow for a 
plugin-like structure? Evidently KDE okular would need to support
this, and I have not asked the KDE devs or investigated how difficult
it would be.

What would be your take in regards to GUI viewers or .pdf handlers
using qpdf as kind of a "plug in"? I do not know how much code 
may have to be modified (I don't really know the internals well
and my C/C++ knowledge is too limited - once it ends up about
memory my brain quits).

Anyway - this is more just a loose idea, please do feel free to disregard
it. It's more meant to get some "feel" for what could be possible, what
could be wanted, if there is any use case (other users) and so forth.

Thanks for reading!

**Comment by jberkenbilt** [comment]
This is a really interesting idea, and it's not that far off from some of what I've been thinking about recently. See #593 for an ongoing discussion. I'm working actively right now on a json interface to qpdf. This will initially just provide an alternative to the qpdf CLI for doing exactly what the CLI does from the library, but an explicit goal of this is to make it possible to write a gui in any language that would dynamically populate a form based on what is available in the current qpdf. The idea would be that you could run a command from the CLI that would spit out a "schema" (not formal like json schema but just something schema-like that would tell you the structure) which could drive an interactive form. This would make it easy to generate a json structure that could passed to qpdf on the CLI or via the library. For your purposes of just adding and removing pages, this might actually be sufficient. Since okular is in C++, it could interface with the library this way if that would be easier, but it wouldn't be necessary.

A hope would be that, once this JSON style of interaction is in place, I can start exposing other library functionality through a json interface which might include the ability to make more general mutations, but that may be a ways off.

Regarding the idea of using qpdf as a plugin, my guess is that this would have a lot more to do with the architecture of okular than with qpdf. I use okular, but I don't know how it's implemented. I'll leave this discussion open so others can chime in or we can throw ideas here. You might take a look at #593, though note that a significant amount of that work has been done. I hope to release the CLI-equivalent json interface as qpdf 10.6 within the next few weeks, then to make some broader changes for qpdf 11. I'm also hoping to implement some more usable C++ API for page manipulation that would make the library much easier to use once I fix a number of page splitting/merging limitations around handling document-level structures.

Bottom line: I like the direction you're thinking in, and some already planned enhancements would make it easier to do this than it is now. Also using qpdf with okular this way is probably more to do with changes in okular than qpdf.
